real_time_control:
  Kp_x: 0.4
  Ki_x: 0.001
  Kd_x: 0.1

  Kp_y: -0.6
  Ki_y: 0.0
  Kd_y: -0.2

  Kp_z: 10.0
  Ki_z: 0.2
  Kd_z: 3.0

  max_thrust: 4.905
  min_thrust: 0.0
  hz: 20  # Frequency in Hz

mid_level_controller:
  # PID gains for thrust, phi, theta
  kp_thrust: 4.5
  ki_thrust: 0.0
  kd_thrust: 2.0

  kp_phi: -0.4
  ki_phi: 0.0
  kd_phi: -0.1

  kp_theta: 0.4
  ki_theta: 0.0
  kd_theta: 0.15

  integral_min: 0.0
  integral_max: 4.905

  min_thrust: 0.0
  max_thrust: 4.905

  dt: 0.05            # Time step for 20 Hz
  
low_level_controller:
  # PID gains for torques
  kp_torque1: 0.02
  ki_torque1: 0.0
  kd_torque1: 0.01

  kp_torque2: 0.02
  ki_torque2: 0.0
  kd_torque2: 0.01

  kp_torque3: 0.02
  ki_torque3: 0.0
  kd_torque3: 0.01

  integral_min: 0.0
  integral_max: 0.3

  dt: 0.01            # Time step for 100 Hz

state_derivative_solver_node:
  # Quadcopter parameters and settings
  m: 0.382            # Mass of the quadcopter (kg)
  I: [0.00062, 0.0, 0.0, 0.0, 0.00062, 0.0, 0.0, 0.0, 0.00124] # Inertia matrix (flattened row-wise)
  g: [0.0, 0.0, -9.81] # Gravity vector (m/s^2)
  dt: 0.002           # Time step for 500 Hz

  tau_thrust: 0.1
  tau_torque: 0.2

  rho: 1.225
  A_rotor: 0.001905
  A_trans: [0.003, 0, 0, 0, 0.003, 0, 0, 0, 0.03]
  r_propeller: 0.0635
  Cd_rot: 0.02
  Cd_trans: [0.08, 0, 0, 0, 0.08, 0, 0, 0, 0.8]
  kt_coeff: 0.0000007
  km_coeff: 0.00000002
  I_rotor: 0.0000215
  l: 0.1
  
  # Impulse force parameters
  impulse_force: [0.0, 0.0, 0.0]  # Force applied as an impulse (Newton)
  force_impulse_start_time: 8.0  # Time when the force impulse starts (seconds)
  force_impulse_duration: 0.5  # Duration of the force impulse (seconds)

  # Impulse torque parameters
  impulse_torque: [0.0, 0.0, 0.0]  # Torque applied as an impulse (Nm)
  torque_impulse_start_time: 15.0  # Time when the torque impulse starts (seconds)
  torque_impulse_duration: 0.05  # Duration of the torque impulse (seconds)